### PR TITLE
fix(helm): Add processing of global helmignore file

### LINKED ISSUE #1674: read helmignore from user's home directory
Git recognizes `~/.gitignore` (in addition to `$repo/.gitignore`) for file patterns that are specific to the user rather than the repo, e.g. editor backup/swap files. `helmignore` should work the same way.

### PR DESCRIPTION
**Pull request by sbezverk**
In addition to processing helmignore file located in the current
directory, this patch adds processing of the global helmignore located
in helm's client default location.

Closes #1674

### CODE REVIEW

**Comment by sbezverk** [comment]
I signed it

File: pkg/chartutil/load.go
Comment (technosophos) [inline pkg/chartutil/load.go]: So there's an unwritten rule in the Helm codebase that nothing inside of `pkg` can assume knowledge of the $HELM_HOME. This is so that other programs can be written using the `pkg` libraries.

When I was talking about this with @adamreese, he had an excellent idea:

- Change `LoadDir()`'s function signature to be `LoadDir(dir string, ignoreFiles ...string) (*chart.Chart, error)`
- Change `Load()`'s signature the same way

Then you can pass in the path to the ignore file whenever you can calculate it. And since we use the `...` notation, that parameter can be omitted. So `LoadDir("foo")` will still work. Then we also don't break backward compatibility.

This will mean fixing a few of the cmd/helm files to call Load or LoadDir with the right args, but @adamreese and I can help track those down. It might even be easier to do that in subsequent pull requests instead of doing it all in one.

File: pkg/chartutil/load.go
Comment (technosophos) [inline pkg/chartutil/load.go]: This function should have a unit test.

**Review by technosophos — CHANGES_REQUESTED** [review]
Overall, this is looking really fabulous. There's one big architectural change we'll unfortunately have to make in order to stay consistent with our rule that `pkg` does not know about `HELM_HOME`, but I don't think the fix will be too bad.

File: pkg/chartutil/load.go
Comment (sbezverk) [inline pkg/chartutil/load.go]: I got the idea of passing list of ignore files to LoadDir and it sounds great but I have a couple of questions:
1. Where do you see ignoreFiles string gets instantiated/populated? at the initialization time? Would be great to have a pointer which function would be the best candidate.
2. Since multiple files can be specified, I would like to use slice of strings.  
Appreciate your reply..

File: pkg/chartutil/load.go
Comment (technosophos) [inline pkg/chartutil/load.go]: What if you only pass the one global ignore file here, and then do this:

```
ifile := filepath.Join(topdir, ignore.HelmIgnore)
if _, err := os.Stat(ifile); err == nil {
    ignoreFiles = append(ignoreFiles ifile)
}       
```

Comment (sbezverk) [inline pkg/chartutil/load.go]: If I could I would prefer not to separate handling global and local helmignore files. It will make the logic more hidden from people looking at the code.

Comment (sbezverk) [inline pkg/chartutil/load.go]: If I could I would prefer not to separate handling global and local helmignore files. It will make the logic more hidden from people looking at the code. If I could add a globally visible function, which will be called by functions using LoadDir, then it might be a bit better. What do you think?

**Comment by technosophos** [comment]
@sbezverk It looks like this PR may actually solve the problem you pointed out above in a way that I had not considered:
https://github.com/kubernetes/helm/pull/2094

Maybe we should hold off on this one until that gets merged, then use the `environment`/`settings` thing there?

**Comment by technosophos** [comment]
Is this PR still moving forward, or is it stalled?

**Pull request closed without merging**
